packages/api/src/store/queue.ts

```diff
@@ -0,0 +1,155 @@
+import sql, { SQLStatement } from "sql-template-strings";
+import { DB } from "./db";
+import logger from "../logger";
+import { BadRequestError } from "./errors";
+import { QueryResult, PoolClient, Pool } from "pg";
+import Table from "./table";
+import { Queue } from "../schema/types";
+import { Notification } from "pg/index";
+import { TableSchema, GetOptions, DBObject, DBLegacyObject } from "./types";
+
+const DEFAULT_SORT = "id ASC";
+
+export default class QueueTable extends Table<Queue> {
+  db: DB;
+  schema: TableSchema;
+  client: PoolClient;
+  listener: PoolClient;
+  channel = "webhook_main";
+  constructor({ db, schema }) {
+    super({ db, schema });
+    this.start();
+  }
+
+  async start() {
+    this.client = await this.db.pool.connect();
+    await this.listen();
+  }
+
+  stop() {
+    try {
+      this.client.release();
+      this.listener.release();
+    } catch (error) {
+      console.log("error releasing pg client", error);
+    }
+  }
+
+  async listen() {
+    this.listener = await this.db.pool.connect();
+    this.listener.query(`LISTEN ${this.channel}`);
+    this.listener.on("notification", async (msg) => {
+      console.log("listener got notification: ", msg.channel);
+      await this.handleMsg(msg);
+    });
+    this.listener.on("error", (error) => {
+      console.error("Msg Queue Listener Error ", error);
+    });
+    this.listener.on("end", () => {
+      console.log("Msq queue listener ended");
+    });
+    this.listener.on("notice", (error) => {
+      console.warn("Msg Queue Listener Notice: ", error);
+    });
+  }
+
+  async handleMsg(msg: Notification) {
+    // emit it here
+    console.log("got NOTIFICATION: ", msg.channel, msg.payload);
+  }
+  // name: string;
+  // constructor({ db, schema }) {
+  //   super({ db, schema });
+  //   this.db = db;
+  //   this.schema = schema;
+  //   this.name = schema.queue;
+  // }
+
+  // get next event in queue
+  async pop(processFunc: Function): Promise<Queue> {
+    let res: QueryResult<DBLegacyObject>;
+    await this.client.query("BEGIN");
```

Comment:
With all the async functions in here I'm a little concerned about what happens when this function gets called a lot. If we call `pop()` twice without waiting for the async, that could cause problems, yeah? Can we add a test something like this to make sure it works okay?

```javascript
for (let i = 0; i < 10; i++) {
  await db.queue.emit({ id: `queue-${i}` });
}

const promises = [];
for (let i = 0; i < 10; i++) {
  promises.push(db.queue.pop());
}

const results = await Promise.all(promises);

// expect results to contain `queue-0`, `queue-1`, etc...

```

My concern is that BEGIN; is going to be called 10 times before anything else happens, right? Because `db.queue.pop()` could run multiple times before any one of the transactions completes. So the transaction might get messed up. Is there a way for us to get a separate transaction object from the pg adapter? Or maybe run everything as part of one big statement?

Reply:
>  Is there a way for us to get a separate transaction object from the pg adapter?

Transaction is per-connection, so every `pop` call should use own connection to the db.

Reply:
`pop` function should take new connection from the pool and use it exclusively for processing one event.